src/waveform/vsyncthread.cpp

```diff
@@ -46,16 +49,70 @@ void VSyncThread::run() {
             m_sinceLastSwap = m_timer.restart();
             m_waitToSwapMicros = 1000;
             usleep(1000);
-        } else { // if (m_vSyncMode == ST_TIMER) {
+        }
+    } else if (m_vSyncMode == ST_PLL) {
+        qint64 offset = 0;
+        qint64 nextSwapMicros = 0;
+        while (m_bDoRendering) {
+            // Use a phase-locked-loop on the QOpenGLWindow::frameSwapped signal
+            // to determine when the vsync occurs
+
+            qint64 pllPhaseOut;
+            qint64 pllDeltaOut;
+            qint64 now;
+
+            {
+                std::scoped_lock lock(m_pllMutex);
+                // last estimated vsync
+                pllPhaseOut = std::llround(m_pllPhaseOut);
+                // estimated frame interval
+                pllDeltaOut = std::llround(m_pllDeltaOut);
+                now = m_pllTimer.elapsed().toIntegerMicros();
+            }
+            if (pllPhaseOut > nextSwapMicros) {
+                nextSwapMicros = pllPhaseOut;
+            }
+            if (nextSwapMicros == pllPhaseOut) {
+                nextSwapMicros += pllDeltaOut;
+            }
+
+            // sleep an integer number of frames extra to approximate the
+            // selected framerate (eg 10,15,20,30)
+            const auto skippedFrames = (m_syncIntervalTimeMicros - pllDeltaOut / 2) / pllDeltaOut;
+            qint64 sleepForSkippedFrames = skippedFrames * pllDeltaOut;
+
+            qint64 sleepUntilSwap = (nextSwapMicros + offset - now) % pllDeltaOut;
+            if (sleepUntilSwap < 0) {
+                sleepUntilSwap += pllDeltaOut;
+            }
+            usleep(sleepUntilSwap + sleepForSkippedFrames);
+
+            m_sinceLastSwap = m_timer.restart();
+            m_waitToSwapMicros = pllDeltaOut + sleepForSkippedFrames;
+
+            // Signal to swap the gl widgets (waveforms, spinnies, vumeters)
+            // and render them for the next swap
+            emit vsyncSwapAndRender();
+            m_semaVsyncSlot.acquire();
+            m_semaVsyncSlot.acquire();
```

Comment:
I think here we've a risk of a race condition between threads, which can increase the number of semaphore ressources. This is because .release() can add resources and we call it twice in vsyncSwapAndRender():
https://doc.qt.io/qt-5/qsemaphore.html#release

Reply:
I don't think that is true. As long as the number of acquires and releases is balanced, and it is.

That said, I agree it looks strange and a single acquire/release will do in this mode, so I will change this.

Reply:
Done